### ISSUE TITLE
Jax compile doesn't release GIL

### ISSUE THREAD

**Issue by chaokunyang** [google/jax#16978]
### Description

Jax compile doesn't release GIL:
![image](https://github.com/google/jax/assets/12445254/dae9000f-975f-41cd-88bf-81cae490abf2)

### What jax/jaxlib version are you using?

jax 0.4.13, jaxlib 0.4.13

### Which accelerator(s) are you using?

CPU

### Additional system info

Linux, Python3,.8

**Comment by jakevdp** [comment]
Hi - thanks for the report. Can you say more about what behavior you're seeing, and how it's different from what you expect? I'm not sure how to interpret a partial screenshot of logs without more context.

**Comment by chaokunyang** [comment]
Sure, we're building a nearly-online computing framework using jax. The computing is complex with 5000~10000 jax operations, which make the XLA compilation superslow, took about 100s, But the actual computing took only less than 10 seconds.

So I made two actions:
- use a separate python thread to compile jax to xla, and let the main thread using jax normal mode to execute the computing. And swithed to jax_compiled when it's finished.
- Use a Python ThreadPoolExecutor to run jax graph concurrently with different inputs.

Seems neither of those action made the program faster.

I sampled several python stack, seems always active on jax_compile, does it mean jax comiple didn't release python GIL?

**Comment by jakevdp** [comment]
Thanks for the details. Would it be possible to construct a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) demonstrating the key features that you're observing? It's hard from just a verbal description of code to actually understand what might be going on when that code runs.

**Comment by hawkinsp** [comment]
JAX compilation should release the GIL, so something else is probably going on. As Jake says, a reproducer would be helpful.

**Comment by chaokunyang** [comment]
OK, I'll try to get a minimal reproduction code here. BTW, is there any other way to make XLA compilation faster?

**Comment by jakevdp** [comment]
> is there any other way to make XLA compilation faster?

Hard to say without an example of what's making it slow. But in general
- if your code is using Python control flow you can make it faster by switching to structured control flow primitives
- if shape changes are triggering frequent recompilations you can change your code to use padded inputs
- if you are using large Python collections (lists, tuples, etc.) you can make compilation faster by switching to keeping your data in native JAX arrays
- etc.

**Comment by chaokunyang** [comment]
Thanks for those suggestions.We've already padding the data to make shape and dtype fixed. It shouldn't be the issue. Python collections are not used either. The jax operations are generated by our framework commputing graph by 50~60 operandas. It's not easy to reproduce it here. I can extract jaxpr here first to see whether the compilation do need so much time. And try to find a way to reproduce the code using jax directly.

**Comment by chaokunyang** [comment]
We'd like to help improve this compilation time

**Comment by jjyyxx** [comment]
@jakevdp Could the document clarify fundamental JAX calls (compilation & execution) that actually release GIL? I found that when using CPU backend, using a `ThreadPoolExecutor` could effectively parallelize my workload (somehow much better than `xla_force_host_platform_device_count+pmap/jit` would do). This maybe indicate JAX execution actually releases GIL, but I cannot find relevant information in documentation.

**Comment by hawkinsp** [comment]
@jjyyxx Unfortunately it's going to be hard to document exactly what releases the GIL. There are dozens or hundreds of entry points into C++, some of which release the GIL, some of which do not. However, compilation and execution are both things that do release the GIL.

**Comment by jjyyxx** [comment]
@hawkinsp Thanks for the clarification! I (and most users, I guess) would only care about GIL for compilation and execution (sorry I did not express that clear enough). I hope this could be mentioned in the documentation, e.g., in https://jax.readthedocs.io/en/latest/concurrency.html.

**Comment by chaokunyang** [comment]
After a detailed test by mocking a graph compilation which took minutes to compile, turns out the jax did released gil when compilation. 
Still not findout why multi thread didn't speed up the jax execution. But it belong to another issue.
Sorry for this misleading issue report.
